Fix more threading issues

mattcaswell opened the pull request:
This fixes a number of threading issues:

1) Ensure the EVP_PKEY operation_cache is appropriately locked

The EVP_PKEY operation_cache caches references to provider side key
objects that have previously been exported for this EVP_PKEY, and their
associated key managers. The cache may be updated from time to time as the
EVP_PKEY is exported to more providers. Since an EVP_PKEY may be shared by
multiple threads simultaneously we must be careful to ensure the cache
updates are locked.

Fixes #13818

2) Don't update rate_limit every time we call ossl_prov_is_running()

Thread santizer was complaining about the unlocked update to the static
variable rate_limit in the FIPS module. In practice this is only important
in the event of the module having gone into the error state. Technically
this access should be locked, but it is only used to limit the rate of
errors that are raised on the error stack. It's not particularly important
if it gets this rate limiting wrong and a lock for this seems over the top.
So we just fix the concurrent access in the normal running state in order
to appease thread sanitizer, and ignore the concurrent access in the error
state.

3) Always ensure we hold ctx->lock when calling CRYPTO_get_ex_data()

Otherwise we can get data races.

4) Avoid races by caching exported ciphers in the init function

TSAN was reporting a race of the exported ciphers cache that we create in
the default and fips providers. This was because we cached it in the query
function rather than the init function, so this would cause a race if multiple
threads queried at the same time. In practice it probably wouldn't make much
difference since different threads should come up with the same answer.

5) Refactor RAND_get0_primary() locking

Make sure we never read or write to dgbl->primary outside of a lock.

I've also extended the multi-threads test. This was primarily with the objective of making tsan spot the unlocked operation_cache issue which was identified in #13818, but it also identified a number of the other issues above.

Finally I've added a CI build to run the threads_test with tsan on.

paulidale commented on providers/fips/self_test.c:
```diff
@@ -359,9 +359,13 @@ int ossl_prov_is_running(void)
-    if (res) {
-        rate_limit = 0;
-    } else if (FIPS_state == FIPS_STATE_ERROR) {
+    if (FIPS_state == FIPS_STATE_ERROR) {
```
`SELF_TEST_post()` only accesses `FIPS_state` while holding the `self_test_lock` lock, even for read.  I'm wondering if the accesses here and in `ossl_set_error_state()` should also have locks.

The semantics of the rate limiting have changes but I don't think not resetting the `rate_limit` is an issue since the module can't exit the error state.

mattcaswell replied:
I've complete reworked the locking in this file to address this issue. It turns out the easiest thing to do was to introduce a new lock for accessing `FIPS_state` rather than reusing `self_test_lock`. My first attempt with just `self_test_lock` was running into some deadlock issues.

t8m commented:
CI failures are relevant.

mattcaswell commented:
Fixups pushed addressing the feedback and  also hopefully addressing the CI failures. The `operation_cache` locking turned out to be insufficient, and I also reworked the locking in the self_test.c file.

Please take a look.

mattcaswell commented:
I have a nasty feeling that the key downgrade code is going to have big threading problems...but I haven't tested it yet and I probably won't bring that in to scope for this PR.

mattcaswell commented:
> I have a nasty feeling that the key downgrade code is going to have big threading problems...but I haven't tested it yet and I probably won't bring that in to scope for this PR.

My suspicions are confirmed: #14020

openssl-machine commented:
This pull request is ready to merge

mattcaswell commented:
Pushed. Thanks.

kaduk commented:
FYI I bisected to this commit that the tests are hanging on my FreeBSD system (in test_encoder_decoder_legacy).
I will try to debug more but there may be lingering issues.

kaduk commented:
> FYI I bisected to this commit that the tests are hanging on my FreeBSD system (in test_encoder_decoder_legacy).
> I will try to debug more but there may be lingering issues.

This is a clear attempt to recursively CRYPTO_thread_write_lock() an EVP_PKEY's lock, which on my system is using pthread_rwlock and (self-)deadlocks.

In test_key(idx==2), the call to test_protected_PEM() from line 535 in turn calls PEM_read_bio_DSAPrivateKey() at line 339, which eventually ends up in evp_pkey_downgrade():

```
1940        if (!CRYPTO_THREAD_write_lock(pk->lock))
(gdb) p pk
$3 = (EVP_PKEY *) 0x800ec6780
```
we then end up in evp_pkey_free_it() and get to evp_keymgmt_util_clear_operation_cache() that tries to lock the already-locked lock:

```
1883                evp_pkey_free_it(*dest);
(gdb) p *dest
$5 = (EVP_PKEY *) 0x800ec6780
(gdb) s
evp_pkey_free_it (x=0x800ec6780) at crypto/evp/p_lib.c:1631
1631        evp_keymgmt_util_clear_operation_cache(x, 1);
(gdb) s
evp_keymgmt_util_clear_operation_cache (pk=0x800ec6780, locking=1)
    at crypto/evp/keymgmt_lib.c:206
207             if (locking && pk->lock != NULL && !CRYPTO_THREAD_write_lock(pk->loc
k))
(gdb) p pk
$6 = (EVP_PKEY *) 0x800ec6780
```

mattcaswell commented:
@kaduk - can you raise this as a separate issue?

kaduk commented:
Yes, this is #14171 (which I filed and then stepped away before realizing I didn't actually cross-link)
